core/amber/src/main/scala/edu/uci/ics/texera/workflow/operators/visualization/wordCloud/WordCloudOpFinalExec.java

```diff
@@ -38,42 +46,79 @@ public void close() {
+            termFreqTuples.add(Tuple.newBuilder().add(
+                    resultSchema,
+                    Arrays.asList(
+                            e.getKey(),
+                            (int) ((e.getValue() - minValue) / (maxValue - minValue) *
+                                    (MAX_FONT_SIZE - MIN_FONT_SIZE) + MIN_FONT_SIZE)
+                    )
+            ).build());
+        }
+        return termFreqTuples;
+    }
+
+    public List<Tuple> calculateResults(List<Tuple> normalizedWordCloudTuples) {
+        List<Tuple> retractions = new ArrayList<>(prevWordCloudTuples);
+        List<Tuple> insertions = new ArrayList<>(normalizedWordCloudTuples);
+
+        retractions.removeAll(normalizedWordCloudTuples);
+        insertions.removeAll(prevWordCloudTuples);
+
+        List<Tuple> results = new ArrayList<>();
+        retractions.forEach(tuple -> results.add(ProgressiveUtils.addRetractionFlag(tuple)));
+        insertions.forEach(tuple -> results.add(ProgressiveUtils.addInsertionFlag(tuple)));
+        return results;
     }
 
     @Override
     public Iterator<Tuple> processTexeraTuple(Either<Tuple, InputExhausted> tuple, int input) {
-       if(tuple.isLeft()) {
-           String term = tuple.left().get().getString(0);
-           int frequency = tuple.left().get().getInt(1);
-           termFreqMap.put(term, termFreqMap.get(term)==null ? frequency : termFreqMap.get(term) + frequency);
-           return JavaConverters.asScalaIterator(Iterators.emptyIterator());
-       }
-       else {
-           double minValue = Double.MAX_VALUE;
-           double maxValue = Double.MIN_VALUE;
-
-           for (Map.Entry<String, Integer> e : termFreqMap.entrySet()) {
-               int frequency = e.getValue();
-               minValue = Math.min(minValue, frequency);
-               maxValue = Math.max(maxValue, frequency);
-           }
-           // normalize the font size for wordcloud js
-           // https://github.com/timdream/wordcloud2.js/issues/53
-           List<Tuple> termFreqTuples = new ArrayList<>();
-           for (Map.Entry<String, Integer> e : termFreqMap.entrySet()) {
-               termFreqTuples.add(Tuple.newBuilder().add(
-                       resultSchema,
-                       Arrays.asList(
-                               e.getKey(),
-                               (int) ((e.getValue() - minValue) / (maxValue - minValue) *
-                                       (this.MAX_FONT_SIZE - this.MIN_FONT_SIZE) + this.MIN_FONT_SIZE)
-                       )
-               ).build());
-           }
-           return JavaConverters.asScalaIterator(termFreqTuples.iterator());
-       }
+        if(tuple.isLeft()) {
+            String term = tuple.left().get().getString(0);
+            int frequency = tuple.left().get().getInt(1);
+            termFreqMap.put(term, termFreqMap.get(term)==null ? frequency : termFreqMap.get(term) + frequency);
+
+            counterSinceLastUpdate++;
```

Comment:
seems like this `counterSinceLastUpdate` is just an indicator of whether there exist some partial outputs that need to be taken care of in the last batch. How about changing it to a boolean?